### ISSUE TITLE
Using NOTIFY to remove need for polling

### ISSUE THREAD

**Issue by alexrudd** [message-db/message-db#31]
Hi, I'm currently evaluating various event stores for a project I'm working on and message-db is really appealing!

Going through the docs it looks like the only way to setup a subscription is to continuously poll the `get_*_messages` procedures and look for changes. I was wondering why [NOTIFY](https://www.postgresql.org/docs/9.0/sql-notify.html) isn't used to broadcast new event writes so that clients can choose to passively listen instead of actively polling. The payload could include the stream name, type, and global position - should be enough for any client to decide if the event is relevant to its subscriptions.

I know very little about postgres, so there might be a good/obvious reason not to do this. Asking out of curiosity :)

**Comment by alexrudd** [comment]
Put together a Go wrapper while exploring Message DB that might be useful to others: https://pkg.go.dev/github.com/alexrudd/gomdb

**Comment by sbellware** [comment]
> Put together a Go wrapper while exploring Message DB that might be useful to others: https://pkg.go.dev/github.com/alexrudd/gomdb

@alexrudd There have been a few people over the past few years who've done some work on Go clients. Feel free to talk about your efforts on the Eventide Slack: https://eventide-project-slack.herokuapp.com/

**Comment by sbellware** [comment]
@alexrudd Using NOTIFY would result in interrupting the polling interval only for streams that are low volume, which is ultimately a self-contradiction.

If a request is made for more messages using one of the message retrieval Postgres server functions, and if that request returns messages, then the next request for more messages should be made immediately after those messages are processed, with no polling interval delay between requests. The best assumption here is that the stream has a steady supply of messages.

The polling should only start when the second request is made and no messages are returned, which indicates that the stream does not have a steady supply of messages.

If NOTIFY is used, it will shorten the amount of time between requests once a consumer has gone into polling.

Let's say that a polling interval is set to 200 milliseconds. And let's say that the subsequent request for messages has resulted in no messages returned, and the consumer starts to poll. This scenario is very literally saying the following: _A stream with message arrival rates of longer than 200 milliseconds should not wait on average 200 milliseconds to check for new messages._ It's not a realistic scenario.

Consider a stream with low expected message arrival rates. Let's say that messages arrive on average once per second, so the polling interval is set to 1000 milliseconds. Again, the expected _and accepted_ average throughput rate is one message per second. So, needing to interrupt the polling for such a low throughput scenario is largely nonsensical.

The only scenario where NOTIFY _is_ valuable is so rare that it precludes Message DB from being an appropriate technology to apply to it:

If a stream has low volume and erratic message arrival rates - and therefore the polling window can be longer than a stream with high volume and high arrival rates - _and_ if messages _must_ be processed with no delay, then NOTIFY should be used to interrupt the polling interval delay.

This kind of scenario signifies something very special: a stream that receives messages infrequently but absolutely must be processed in near real time. This kind of use case usually only occurs in life-critical applications, and in such scenarios, other technologies are usually used - usually technologies that are closer to the metal and more tightly integrated at the system level.

For the vast majority of business applications for which Message DB is purposed, a polling interval tuned to a stream's expected message arrival rate is perfectly practicable. In practice, the use of NOTIFY is excessive for the vast majority of scenarios.

It _would_ be appropriate to use NOTIFY for critical real time signals processing scenarios, like machine control systems. But they, it would be reasonable to ask whether Postgres is the right transport mechanism in such scenarios, as opposed to raw network signals.

And lastly, even if a scenario existed where something like a 200 millisecond polling interval needed to be interrupted, consider the round trip cycle time of sending the NOTIFY signal, the client receiving and processing the signal, interrupting the client's polling mechanism and reseting its delay, then issuing the message retrieval query. _Some_ amount of time would be saved, but it's worth questioning whether enough time would be saved to matter. In such a case, it might be better to re-turn the polling interval to something lower to reflect the _actual_ expected average arrival rate of messages.

In the end, NOTIFY _in practice_ doesn't solve a real problem that is common enough to matter for the scenarios for which Message DB is intended.

That said, Message DB is all just Postgres server functions... it wouldn't be too much of a stretch for a user to fork the implementation and add support for NOTIFY. But again, in practice no users have found it necessary, not even users implementing financial apps. But then again, Message DB wouldn't be an appropriate technology to build the world's fastest Bitcoin miner (even though it has been used in non-real time crypto scenarios).

**Comment by alexrudd** [comment]
Hi @sbellware, I really appreciate your detailed response! I hadn't considered the advantages of polling to be quite so absolute but this has thoroughly changed my mind. I had implemented a exponential back-off strategy in my client for polling, I may now also look at writing a strategy to optimise polling frequency based on message delivery rate.

I have also joined the Eventide slack so will share my progress there :) Consider this issue done, though leaving it open might be useful for those that follow.

Thanks again!
